File: aoc-2023/rust/src/day1.rs
use std::{fs::read_to_string, collections::HashMap};

fn parse_input() -> String {
    read_to_string("../input/day1.txt")
        .unwrap()
}

fn part_one(input: &String) {
    let x = input
        .split("\n")
        .map(|x| x.chars())
        .map(|x| x.filter(|x| x.is_numeric()))
        .map(|x| x.collect::<String>())
        .map(|x| x[..1].to_string() + &x[x.len()-1..].to_string())
        .map(|x| x.parse::<u32>().unwrap())
        .sum::<u32>();
    println!("{:?}", x)
}

const DIGITS: [&'static str; 9] = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];

fn find_digit(x: &str, rev: bool, digits: &HashMap<&str, String>) -> String {
    let n = x.len() - 1;
    for i in 0..=n {
        let slice = if rev { &x[n-i ..] } else { &x[..i] };
        for (i, digit) in DIGITS.iter().enumerate() {
            let num = (i+1).to_string();
            if slice.contains(digit) {
                return digits.get(digit).unwrap().to_string();
            } else if slice.contains(&num) {
                return num;
            }
        }
    }
    String::new()
}

fn part_two(input: &String) {
    let mut digits = HashMap::new();
    for (i, digit) in DIGITS.iter().enumerate() {
        digits.insert(*digit, (i+1).to_string());
    }
    let x = input
        .split("\n")
        .map(|x| {
            let mut first = find_digit(&x, false, &digits);
            let last = find_digit(&x, true, &digits);
            if first.is_empty() { first = last.clone() };
            first + &last
        })
        .map(|x| x.parse::<u32>().unwrap())
        .sum::<u32>();
    println!("{:?}", x)
}

pub fn main() {
    let input = &parse_input();
    part_one(input);
    part_two(input);
}

File: aoc-2023/python/day1.py
lines = [line.rstrip() for line in open('./input/day1.txt').readlines()]

def part_one():
    res = 0
    for line in lines:
        digits = [x for x in line if x.isdigit()]
        a, b = digits[0], digits[-1]
        res += int(a+b)
    print(res)

numbers = {
    'zero': '0',
    'one': '1',
    'two': '2',
    'three': '3', 
    'four': '4',
    'five': '5', 
    'six': '6', 
    'seven': '7',
    'eight': '8', 
    'nine': '9'
}
digits = list('0123456789') + list(numbers.keys())

def find_digits(line: str):
    first_index, last_index = len(line), -1
    first_digit, last_digit = '', ''
    for digit in digits:
        i = line.find(digit)
        if i > -1:
            if i < first_index:
                first_index = i
                if digit in numbers.keys():
                    first_digit = numbers[digit]
                else:
                    first_digit = digit
        i = line.rfind(digit)
        if i > -1:
            if i > last_index:
                last_index = i
                if digit in numbers.keys():
                    last_digit = numbers[digit]
                else:
                    last_digit = digit
    return first_digit + last_digit

def part_two():
    res = sum([int(find_digits(line)) for line in lines])
    print(res)
    
# part_one()
part_two()